Python/optimizer.c

```diff
@@ -0,0 +1,245 @@
+        code->co_executors = new;
+    }
+    assert(size < code->co_executors->capacity);
+    code->co_executors->size++;
+    return size;
+}
+
+static int
+insert_executor(PyCodeObject *code, int offset, _PyExecutorObject *executor)
+{
+    _Py_CODEUNIT original = _PyCode_CODE(code)[offset];
+    if (original.op.code == ENTER_EXECUTOR) {
+        _PyExecutorObject *old = code->co_executors->executors[original.op.arg];
+        executor->vm_data.opcode = old->vm_data.opcode;
+        executor->vm_data.oparg = old->vm_data.oparg;
+        old->vm_data.opcode = 0;
+        Py_INCREF(executor);
+        code->co_executors->executors[original.op.arg] = executor;
+        Py_DECREF(old);
+    }
+    else {
+        int32_t index = get_next_free_in_executor_array(code);
+        if (index < 0) {
+            return -1;
+        }
+        Py_INCREF(executor);
+        executor->vm_data.opcode = original.op.code;
+        executor->vm_data.oparg = original.op.arg;
+        code->co_executors->executors[index] = executor;
+        assert(index < 256);
+        _PyCode_CODE(code)[offset].op.code = ENTER_EXECUTOR;
+        _PyCode_CODE(code)[offset].op.arg = index;
+    }
+    return 0;
+}
+
+int
+PyUnstable_Replace_Executor(PyCodeObject *code, int offset, _PyExecutorObject *new)
+{
+
+    _Py_CODEUNIT original = _PyCode_CODE(code)[offset];
+    if (original.op.code != ENTER_EXECUTOR) {
+        PyErr_Format(PyExc_ValueError, "No executor to replace");
+        return -1;
+    }
+    return insert_executor(code, offset, new);
+}
+
+static _PyExecutorObject *
+error_optimize(
+    _PyOptimizerObject* self,
+    PyCodeObject *code,
+    _Py_CODEUNIT *instr)
+{
+    PyErr_Format(PyExc_SystemError, "Should never call error_optimize");
+    return NULL;
+}
+
+static PyTypeObject DefaultOptimizer_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "noop_optimizer",
+    .tp_basicsize = sizeof(_PyOptimizerObject),
+    .tp_itemsize = 0,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_DISALLOW_INSTANTIATION,
+};
+
+_PyOptimizerObject _PyOptimizer_Default = {
+    PyObject_HEAD_INIT(&DefaultOptimizer_Type)
+    .optimize = error_optimize,
+    .resume_threshold = UINT16_MAX,
+    .backedge_threshold = UINT16_MAX,
+};
+
+_PyOptimizerObject *
+PyUnstable_GetOptimizer(void)
+{
+    PyInterpreterState *interp = PyInterpreterState_Get();
+    if (interp->optimizer == &_PyOptimizer_Default) {
+        return NULL;
+    }
+    Py_INCREF(interp->optimizer);
+    return interp->optimizer;
+}
+
+void
+PyUnstable_SetOptimizer(_PyOptimizerObject *optimizer)
+{
+    PyInterpreterState *interp = PyInterpreterState_Get();
+    if (optimizer == NULL) {
+        optimizer = &_PyOptimizer_Default;
+    }
+    _PyOptimizerObject *old = interp->optimizer;
+    Py_INCREF(optimizer);
+    interp->optimizer = optimizer;
+    interp->optimizer_backedge_threshold = optimizer->backedge_threshold;
+    interp->optimizer_resume_threshold = optimizer->resume_threshold;
```

Comment:
Should we shift by `OPTIMIZER_BITS_IN_COUNTER` here so that optimizers don't have to always specify thresholds shifted by that amount?

If the value of `OPTIMIZER_BITS_IN_COUNTER` is implicitly hardcoded into the thresholds of every optimizer, it makes it harder to ever change that value.

OTOH I guess optimizers already have to know how many bits are available for their use, and if they ever dynamically adjust their thresholds, they also have to know how to correctly update the interpreter cached thresholds. So maybe it's clearer to just always have thresholds shifted everywhere.

Reply:
We don't want to do the shift, because it prevents optimizers specifying "always off" by setting the value to `UINT16_MAX`.
It is a bit clunky, but this is all "unstable", so we can see how this works out.